File: backend/expenses/src/main/java/onosoft/adapters/driven/expense/dto/ExpenseDto.java
package onosoft.adapters.driven.expense.dto;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.validation.constraints.Size;
import lombok.*;
import lombok.experimental.SuperBuilder;
import onosoft.adapters.driven.commons.money.MoneyDto;
import onosoft.domain.model.PaymentType;

@Getter
@EqualsAndHashCode
@NoArgsConstructor
@ToString
@SuperBuilder
public class ExpenseDto {
    @NonNull
    @Size(max = 30)
    protected String recipient;

    @NonNull
    @Size(max = 30)
    protected String purpose;

    @NonNull
    protected MoneyDto amount;

    @NonNull
    @JsonFormat(pattern="yyyy-MM-dd")
    protected String accruedDate;

    @Builder.Default
    protected final PaymentType paymentType = PaymentType.Unknown;

    @JsonFormat(pattern="yyyy-MM-dd")
    protected String paymentTargetDate;

    @JsonProperty("isInvoiced")
    protected boolean invoiced;
}

File: frontend/expenses/features/expenses/features/assign/api/ExpenseDto.ts
import { Expense } from "@/features/expenses/types/Expense";

export namespace ExpenseDto {
  export type Type = Omit<
    Expense.Type,
    "accruedDate" | "paymentTargetDate" | "paymentActualDate"
  > & {
    accruedDate: string;
    paymentTargetDate: string;
    paymentActualDate: string;
  };

  export function of(domain: Expense.Type): Type {
    return {
      ...domain,
      amount: { ...domain.amount },
      accruedDate: domain.accruedDate.toLocaleDateString("sv"),
      paymentTargetDate: domain.paymentTargetDate.toLocaleDateString("sv"),
      paymentActualDate: domain.paymentActualDate
        ? domain.paymentActualDate.toLocaleDateString("sv")
        : "",
    };
  }
}
